src/mscorlib/src/System/Buffer.cs

```diff
@@ -425,10 +437,213 @@ internal unsafe static void Memmove(byte* dest, byte* src, nuint len)
 #endif
             return;
 
-            PInvoke:
+PInvoke:
             _Memmove(dest, src, len);
         }
-        
+
+        // This method has different signature for x64 and other platforms and is done for performance reasons.
+        internal static void Memmove(ByReference<byte> dest, ByReference<byte> src, nuint len)
+        {
+#if AMD64 || (BIT32 && !ARM)
+            const nuint CopyThreshold = 2048;
+#elif ARM64
+#if PLATFORM_WINDOWS
+            // Determined optimal value for Windows.
+            // https://github.com/dotnet/coreclr/issues/13843
+            const nuint CopyThreshold = UInt64.MaxValue;
+#else // PLATFORM_WINDOWS
+            // Managed code is currently faster than glibc unoptimized memmove
+            // TODO-ARM64-UNIX-OPT revisit when glibc optimized memmove is in Linux distros
+            // https://github.com/dotnet/coreclr/issues/13844
+            const nuint CopyThreshold = UInt64.MaxValue;
+#endif // PLATFORM_WINDOWS
+#else
+            const nuint CopyThreshold = 512;
+#endif // AMD64 || (BIT32 && !ARM)
+
+            ref byte srcEnd = ref Unsafe.Add(ref src.Value, (IntPtr)len);
+            ref byte destEnd = ref Unsafe.Add(ref dest.Value, (IntPtr)len);
+
+            // P/Invoke into the native version when the buffers are overlapping.
+            //
+            // Buffers don't overlap if the end of one buffer occurs before (or at) the start of the other buffer.
+            // if ((destEnd <= src) || (srcEnd <= dest)) { /* buffers don't overlap */ }
+            // if (!(destEnd > src) || !(srcEnd > dest)) { /* buffers don't overlap */ }
+            // if (!(src < destEnd) || !(dest < srcEnd)) { /* buffers don't overlap */ }
+            // if (!(!(src < destEnd) || !(dest < srcEnd))) { /* buffers overlap */ }
+            // Finally, apply DeMorgan's theorem:
+            // if ((src < destEnd) && (dest < srcEnd)) { /* buffers overlap */ }
+
+            if (Unsafe.IsBelow(ref src.Value, ref destEnd) && Unsafe.IsBelow(ref dest.Value, ref srcEnd))
```

Comment:
This line is incorrect and will be reverted back to its original implementation (or higher-performance equivalent).